### PR TITLE
Migrate components/playback functions to ES6 modules

### PR DESCRIPTION
**Pull request by Delgan**
_**Disclaimer**: this commit has been automated by a script._

**Changes**
Looking at the instructions in #868 it seems to me that much of the work can be automated. I made an experimental script to convert a file to an ES6 module.

Of course, this implies that we have to double-check the commit very carefully. But I suppose it can still save a lot of time. I'm just trying to automate the boring stuff with Python. :)

The script is also capable of converting a function object with `prototype` to a proper ES6 `class`. Unfortunately, it caused errors in other files that import these classes. So I might need help to fix this issue.

See `diff` with white-spaces ignored here: [Files changed](https://github.com/jellyfin/jellyfin-web/pull/1284/files?w=1)

What do you think of such automated changes and should I continue applying it to others files?
 
<details>
<summary>Script used</summary>

```python
#!/usr/bin/env python3
import argparse
import itertools
import json
import os
import re
import sys
import textwrap
import traceback


def convert_to_es6_module(module, package_path):
    with open(module) as file:
        data = file.read()

    match = re.match(
        r"\s*define\(\[(?P<imports>[^\]]*)\]\s*,\s*function\s*\((?P<names>[^\)]*)\)\s*{"
        r"\s*(?:[\"']use strict[\"'];)?"
        r"(?P<content>[\S\s]*)"
        r"}\);",
        data,
    )

    if match is None:
        raise ValueError("Pattern not found")

    groups = match.groupdict()
    imports, names, content, exported = groups["imports"], groups["names"], groups["content"], None

    imports = re.findall(r"'([^']+)'", imports)
    names = re.findall(r"\w+", names)

    match_exported = re.fullmatch(
        r"(?P<content>[\S\s]*)\breturn\s+(?P<exported>\w+|{[^}]*})\s*;\s*", content
    )

    if match_exported is not None:
        groups = match_exported.groupdict()
        content, exported = groups["content"], groups["exported"]

        if "{" in exported:
            export_multiple = True
            count = exported.count(":")
            exported = [
                [r.strip() for r in ret.split(":")]
                for ret in re.findall(r"\w+\s*:\s*\w+", exported)
            ]
            if len(exported) != count:
                raise ValueError("Could not parse all exported functions")
        else:
            export_multiple = False
            name = exported.strip()
            exported = [[name, name]]

    if exported is not None:
        for _, value in exported:
            reg = r"\bfunction\s+%s\s*\(" % re.escape(value)

            match = re.search(reg, content)

            if match is None:
                raise ValueError(f"Can't auto-update function: {value}")

            content = re.sub(reg, f"export function {value}(", content, 1)

    method_header = r"\n(%s)(%s)\.prototype\.(%s)\s*=\s*function\s*\(([^)]*)\)"

    classes = {m[1] for m in re.findall(method_header % (r"[ \t]*", "\w+", r"\w+"), content)}

    for class_name in classes:
        reg = re.escape(class_name)
        methods = ""

        for match in re.finditer(method_header % (r"[ \t]*", reg, r"\w+"), content):
            indentation, _, name, args = match.groups()
            method_body = r"\s*{([\s\S]*?)\n%s};" % indentation
            method_pattern = method_header % (indentation, reg, re.escape(name)) + method_body
            method_match = re.search(method_pattern, content)

            if not method_match:
                raise ValueError(f"Can't update method '{name}' for class: {class_name}")

            *_, body = method_match.groups()
            content = content.replace(method_match.group(0), "", 1)
            methods += "%s%s(%s) {%s\n%s}\n\n" % (indentation, name, args, body, indentation)

        constructor_header = r"\n(%s)(export\s+)?function\s+{}\s*\(([^)]*)\)".format(reg)

        constructor_match = re.search(constructor_header % r"[ \t]*", content)

        if not constructor_match:
            raise ValueError(f"Constructor not found for: {class_name}")

        indentation, export, args = constructor_match.groups()
        export = "export " if export else ""
        constructor_body = r"\s*{([\s\S]*?)\n%s}" % indentation

        constructor_match = re.search(constructor_header % indentation + constructor_body, content)

        if not constructor_match:
            raise ValueError(f"Constructor body not found for: {class_name}")

        *_, body = constructor_match.groups()

        constructor = "%sconstructor(%s) {%s\n%s}" % (indentation, args, body, indentation)

        repl = "\n%sclass %s {\n\n%s\n\n%s\n}" % (export, class_name, constructor, methods)
        content = content.replace(constructor_match.group(0), textwrap.indent(repl, indentation), 1)

        if re.search("\b%s\.prototype\.\w+\s*=\s*function" % reg, content) is not None:
            raise ValueError(f"Some method could not be updated for class: {class_name}")

    content = re.sub(r"\n\n[\n]+", "\n\n", content)

    output = ""

    for import_, name in itertools.zip_longest(imports, names):
        if name is None:
            output += f"import '{import_}';\n"
        else:
            output += f"import {name} from '{import_}';\n"

    if imports:
        output += "\n"

    output += textwrap.dedent(content).strip() + "\n"

    if exported is not None:
        if not export_multiple:
            output += f"\nexport default {exported[0][0]};"
        else:
            output += "\nexport default {\n"
            for key, value in exported:
                output += f"    {key}: {value},\n"
            output = output.rstrip(",\n") + "\n"
            output += "};"
        output += "\n"

    returns = re.findall(r"\nreturn\s+", output)

    if len(returns) > 1:
        raise ValueError("Some functions could not be exported (found multiple global 'return')")
    elif len(returns) == 1:
        output = re.sub(r"\nreturn\s+", "\nexport default ", output, 1)

    with open(module, "w") as f:
        f.write(output)

    if package_path is not None:
        path = os.path.abspath(module)
        if "/jellyfin-web/" not in path:
            raise ValueError("Module is not part of jellyfin-web, can't add it to package.json")
        relative_path = path.split("/jellyfin-web/")[-1]

        with open(package_path) as p:
            data = json.load(p)

        data["babel"]["overrides"][0]["test"].append(relative_path)
        data["babel"]["overrides"][0]["test"].sort()

        with open(package_path, "w") as p:
            p.write(json.dumps(data, indent=2) + "\n")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("file", help="List of files to convert to ES6 modules", nargs="+")
    parser.add_argument("--package", help="Path to the json package file")
    parser.add_argument("--verbose", help="Display full traceback on error", action="store_true")
    args = parser.parse_args()

    for file in args.file:
        name = os.path.basename(file)
        try:
            convert_to_es6_module(file, args.package)
        except Exception as e:
            print(f"\x1b[31m✗\x1b[0m \x1b[1m{name}\x1b[0m: {str(e)}")
            if args.verbose:
                traceback.print_exc()
        else:
            print(f"\x1b[32m✓\x1b[0m \x1b[1m{name}\x1b[0m")
```

</details>

**Issues**
Part of #868

### CODE REVIEW

File: src/components/playback/playerSelectionMenu.js [hunk 1]
```diff
@@ -1,320 +1,325 @@
...
+    }
+}
 
-        if (currentPlayerInfo) {
-            if (!currentPlayerInfo.isLocalPlayer) {
-                showActivePlayerMenu(currentPlayerInfo);
-                return;
-            }
-        }
+export function showPlayerSelection(button) {
```
Comment (dmitrylyzo) [inline src/components/playback/playerSelectionMenu.js:86]: Taking into account `show: showPlayerSelection` below, you need to rename function to `show`.

File: src/components/playback/playersettingsmenu.js [hunk 2]
```diff
@@ -1,270 +1,274 @@
-define(['connectionManager', 'actionsheet', 'datetime', 'playbackManager', 'globalize', 'appSettings', 'qualityoptions'], function (connectionManager, actionsheet, datetime, playbackManager, globalize, appSettings, qualityoptions) {
-    'use strict';
-
-    function showQualityMenu(player, btn) {
+import connectionManager from 'connectionManager';
+import actionsheet from 'actionsheet';
+import datetime from 'datetime';
```
Comment (sarab97) [inline src/components/playback/playersettingsmenu.js:3]: this line can be removed as i dont see this being used anywhere

Comment (sarab97) [inline src/components/playback/playersettingsmenu.js:3]:
```suggestion
```
this line can be removed as datetim isnt being used.

File: src/components/playback/playersettingsmenu.js [hunk 3]
```diff
@@ -1,270 +1,274 @@
...
+import playbackManager from 'playbackManager';
+import globalize from 'globalize';
+import appSettings from 'appSettings';
```
Comment (sarab97) [inline src/components/playback/playersettingsmenu.js:6]:
```suggestion
```
same here

File: src/components/playback/brightnessosd.js [hunk 4]
```diff
@@ -1,171 +1,173 @@
...
-            }
-        }
-        if (progressElement) {
-            progressElement.style.width = (brightness || 0) + '%';
+}
+
+function setIcon(iconElement, icon) {
+    iconElement.classList.remove('brightness_high');
+    iconElement.classList.remove('brightness_medium');
+    iconElement.classList.remove('brightness_low');
```
Comment (sarab97) [inline src/components/playback/brightnessosd.js:108]:
```suggestion
    iconElement.classList.remove('brightness_high', 'brightness_medium', 'brightness_low');
```
slightly better looking and minified :)

Comment (dmitrylyzo) [reply src/components/playback/brightnessosd.js:108]: Agree.
FYI, there is similar code elsewhere (#898). All of them can be changed in this way.

File: src/components/playback/playerSelectionMenu.js [hunk 5]
```diff
@@ -1,320 +1,325 @@
...
-            return target.user.Name;
-        }
+    if (target.user) {
 
-        return null;
+        return target.user.Name;
     }
 
-    function getIcon(target) {
+    return null;
```
Comment (sarab97) [inline src/components/playback/playerSelectionMenu.js:48]:
```suggestion
return target.user ? target.user.Name : null;
```
this complete function can be donel like this

File: src/components/playback/brightnessosd.js [hunk 6]
```diff
@@ -1,171 +1,173 @@
...
+import 'material-icons';
 
-    var currentPlayer;
-    var osdElement;
-    var iconElement;
-    var progressElement;
+var currentPlayer;
+var osdElement;
+var iconElement;
+var progressElement;
```
Comment (sarab97) [inline src/components/playback/brightnessosd.js:11]: arent we supposed to avoid globals.
We also need to get rid of these

**Comment by dkanada** [comment]
Changing everything to `const` and `let` isn't nearly as important though. We only need to remove requirejs to get a proper build system working so that should be the priority over nice additions like new language features.

**Comment by Delgan** [comment]
Thanks for your comments.

There is some stuff that I can certainly automate with the script (remove unused import, merge classList remove/add) but others that I can't (replace var with let/const, refactor functions and globals).

I was under the impression that #868 was only about migrating the import/export process of modules, not actually fully modernize the code. That's why my PR is limited to basic changes. I hope adding ES6 features can be done in another incremental step.

**Comment by sarab97** [comment]
> Thanks for your comments.
> 
> There is some stuff that I can certainly automate with the script (remove unused import, merge classList remove/add) but others that I can't (replace var with let/const, refactor functions and globals).
> 
> I was under the impression that #868 was only about migrating the import/export process of modules, not actually fully modernize the code. That's why my PR is limited to basic changes. I hope adding ES6 features can be done in another incremental step.

the importance and priority is given to migrations but in the process if somethings comes up some small change with improvements which can be done we do it. Still a good work on python script. And a good work by your python script too.

**Comment by MrTimscampi** [comment]
I agree with @dkanada that the primary process is axing RequireJS from the code so we can avoid relying on my hacky Gulp.js config to get transpiling and such.

We've taken the opportunity to refactor while doing this, because if we're in there, why not take 10 minutes to do more stuff (which often ends up being more than 10 minutes, but still).

Thing is: most refactor can be done later. We can probably get it over quicker by running this over files and fixing them up manually if they don't work.

Then the refactor will be much easier, as we'll be able to get rid of aliases and such, that make it harder than needed to develop for this.

**Comment by MrTimscampi** [comment]
To add to this: while we've encouraged refactors during migration, ultimately, it's up to every contributor :) We only ask that stuff works for it to be merged, the rest is all bonus :)

**Comment by MrTimscampi** [comment]
@Delgan If you can fix the conflicts, this is ready for a merge :)

**Comment by Delgan** [comment]
@MrTimscampi Done. :+1:

**Pull request merged**